File: soccer-form-ai/js/ballDetection.js
/* ballDetection.js — finds the ball in a clip using MediaPipe's ObjectDetector.
 *
 * Why this is a module and not a single detect() call: one frame's raw output
 * is not trustworthy for this job. The COCO "sports ball" class also fires on
 * round-ish background clutter (in testing, on a stretched patch of grass that
 * scored 0.33), and the one frame we care about most — contact — is exactly
 * where the ball is fastest, most motion-blurred, and most likely to be missed
 * outright. Either failure would quietly corrupt a metric.
 *
 * So we read the WHOLE clip and pick the single most temporally coherent track
 * through it. A real ball moves a short, smooth distance frame to frame; a
 * false positive appears once somewhere unrelated and vanishes. Frames the
 * detector missed inside an otherwise good track are interpolated and flagged
 * as such; frames outside it stay null rather than being extrapolated.
 *
 * Every gate below is expressed relative to the player's own torso length,
 * measured by the pose model on the same frame — so, as everywhere else in
 * this app, there are no hard-coded pixel distances.
 *
 * API confirmed against the published types for @mediapipe/tasks-vision@0.10.35:
 *   ObjectDetector.createFromOptions(fileset, options): Promise<ObjectDetector>
 *   detectForVideo(videoFrame, timestamp, opts?): DetectionResult
 *   DetectionResult.detections[].boundingBox = {originX, originY, width, height}
 *   DetectionResult.detections[].categories[] = {categoryName, score, ...}
 * Bounding boxes come back in PIXELS of the frame passed in, not normalized.
 */

import { FilesetResolver, ObjectDetector }
  from 'https://cdn.jsdelivr.net/npm/@mediapipe/tasks-vision@0.10.35/vision_bundle.mjs';

const WASM_PATH = 'https://cdn.jsdelivr.net/npm/@mediapipe/tasks-vision@0.10.35/wasm';
// EfficientDet-Lite0, uint8 quantized: 4.5MB, versus 13.8MB for the float32
// build of the same model. On a phone that download difference matters more
// than the accuracy difference does, because the temporal track below is what
// actually carries reliability here, not any single frame's confidence.
const MODEL_URL =
  'https://storage.googleapis.com/mediapipe-tasks/object_detector/efficientdet_lite0_uint8.tflite';

const BALL_CATEGORY = 'sports ball';   // the COCO class a football comes back as
const SCORE_FLOOR = 0.2;               // below this the track logic can't rescue it anyway

/* Plausibility gates — all scale-free. */
// A ball's bounding box is near-square from any angle. Long thin boxes are
// line markings, shins, or smeared grass. Kept loose because a ball moving
// fast does smear into an oval.
const MIN_ROUNDNESS = 0.5;
// A size-5 ball is ~22cm across; hip-to-shoulder is ~50cm, so a ball is
// ~0.45 torso lengths wide. The band is generous because the ball can sit
// nearer or further from the camera than the player.
const MIN_DIAM_TORSO = 0.18;
const MAX_DIAM_TORSO = 0.95;

/* Track-selection costs (see bestTrack). */
// Cost of declaring "no ball in this frame". Must sit above the emission cost
// of a plausible detection (1 - score, so ~0.3 at score 0.7) or the track
// would rather see nothing than accept a real ball.
const MISS_COST = 0.8;
const STEP_WEIGHT = 0.6;
// A struck ball leaves at ~25m/s; at 30fps that is ~0.8m, or ~1.7 torso
// lengths, in one frame. Displacement beyond this is capped rather than
// rejected, so a genuinely fast ball is penalised but still reachable.
const MAX_STEP_TORSO = 2.5;

let detector = null;
let backend = null;
let initPromise = null;

// detectForVideo needs strictly increasing timestamps for the life of the
// instance, same constraint as the pose landmarker — and for the same reason
// (phases are separate clips, picked in any order) we drive it off our own
// counter instead of real video time. This one is independent of the pose
// clock; they are different task instances and do not share state.
let clockMs = 0;

function build(vision, delegate) {
  return ObjectDetector.createFromOptions(vision, {
    baseOptions: { modelAssetPath: MODEL_URL, delegate },
    runningMode: 'VIDEO',
    // Ask the model for nothing else. Cheaper than filtering after the fact,
    // and it means a stray "person" or "tennis racket" can never reach us.
    categoryAllowlist: [BALL_CATEGORY],
    scoreThreshold: SCORE_FLOOR,
    maxResults: 8
  });
}

/** Idempotent init. Prefers the GPU delegate, falls back to CPU. */
export function initBall() {
  if (initPromise) return initPromise;
  initPromise = (async () => {
    const vision = await FilesetResolver.forVisionTasks(WASM_PATH);
    try {
      detector = await build(vision, 'GPU');
      backend = 'GPU';
    } catch (err) {
      console.warn('Ball detector: GPU delegate unavailable, falling back to CPU:', err);
      detector = await build(vision, 'CPU');
      backend = 'CPU';
    }
    return { backend };
  })();
  return initPromise;
}

export const getBallBackend = () => backend;

/**
 * Find the ball across one ordered clip.
 *
 * @param {Array<{canvas:HTMLCanvasElement, time:number}>} items clip frames, ascending in time
 * @param {Array<number|null>} scales torso length in px for each frame (null where
 *        no pose was found); used to size- and motion-gate candidates
 * @returns {Array<{x,y,r,score,interpolated}|null>} one entry per input frame,
 *          in that frame's pixel space. null = no trustworthy ball there.
 */
export async function detectBallSequence(items, scales = []) {
  if (!detector) await initBall();

  const filled = fillScales(scales, items.length);
  const perFrame = [];
  let prevTime = null;
  for (let i = 0; i < items.length; i++) {
    const { canvas, time } = items[i];
    const deltaMs = prevTime == null ? 33 : Math.max(1, Math.round((time - prevTime) * 1000));
    clockMs += deltaMs;
    prevTime = time;
    perFrame.push(candidates(detector.detectForVideo(canvas, clockMs), filled[i]));
  }
  clockMs += 200; // gap so the next clip never collides with this one

  return fillGaps(bestTrack(perFrame, filled));
}

/**
 * Torso length is the yardstick for every gate here, but the pose model may
 * have failed on some frames of the clip. Borrow the median of the frames it
 * did read — the player's size does not meaningfully change within ~0.3s.
 */
function fillScales(scales, n) {
  const known = [];
  for (let i = 0; i < n; i++) {
    const s = scales[i];
    if (typeof s === 'number' && isFinite(s) && s > 0) known.push(s);
  }
  const median = known.length
    ? known.slice().sort((a, b) => a - b)[Math.floor(known.length / 2)]
    : 0;
  return Array.from({ length: n }, (_, i) => {
    const s = scales[i];
    return (typeof s === 'number' && isFinite(s) && s > 0) ? s : median;
  });
}

/** Detections from one frame, reduced to plausible ball candidates. */
function candidates(result, scale) {
  const out = [];
  for (const d of result?.detections ?? []) {
    const b = d.boundingBox;
    if (!b || b.width <= 0 || b.height <= 0) continue;
    if (Math.min(b.width, b.height) / Math.max(b.width, b.height) < MIN_ROUNDNESS) continue;

    const diam = (b.width + b.height) / 2;
    // Only size-gate when we actually know how big the player is; with no
    // pose on any frame of the clip we would be inventing a reference.
    if (scale > 0) {
      const rel = diam / scale;
      if (rel < MIN_DIAM_TORSO || rel > MAX_DIAM_TORSO) continue;
    }

    out.push({
      x: b.originX + b.width / 2,
      y: b.originY + b.height / 2,
      r: diam / 2,
      score: d.categories?.[0]?.score ?? 0
    });
  }
  return out;
}

/**
 * Pick the most coherent path through the per-frame candidates (Viterbi).
 *
 * Each frame's states are its candidates plus a "not visible here" state, so
 * the track can survive the ball being missed mid-clip without being forced
 * onto whatever junk the detector did return. Cost of a state is its
 * unconfidence (1 - score); cost of moving between two real candidates is how
 * far the ball would have had to travel, in torso lengths. A false positive is
 * therefore only chosen if nothing more coherent exists — and a lone one, far
 * from everything else, loses to the miss state.
 */
function bestTrack(perFrame, scales) {
  const n = perFrame.length;
  if (!n) return [];

  const MISS = -1;
  const statesAt = i => [...perFrame[i].map((_, k) => k), MISS];
  const emission = (i, k) => k === MISS ? MISS_COST : 1 - perFrame[i][k].score;
  const transition = (i, from, to) => {
    // Nothing to compare against across a gap — moving in or out of the miss
    // state is free, and paying for it twice would just bias against gaps.
    if (from === MISS || to === MISS) return 0;
    const a = perFrame[i - 1][from], b = perFrame[i][to];
    const s = scales[i] || scales[i - 1] || 0;
    if (!s) return 0;
    return STEP_WEIGHT * Math.min(Math.hypot(a.x - b.x, a.y - b.y) / s, MAX_STEP_TORSO);
  };

  let states = statesAt(0);
  let costs = states.map(k => emission(0, k));
  let paths = states.map(k => [k]);

  for (let i = 1; i < n; i++) {
    const next = statesAt(i);
    const nextCosts = [], nextPaths = [];
    for (const to of next) {
      let best = Infinity, bestJ = 0;
      states.forEach((from, j) => {
        const c = costs[j] + transition(i, from, to);
        if (c < best) { best = c; bestJ = j; }
      });
      nextCosts.push(best + emission(i, to));
      nextPaths.push([...paths[bestJ], to]);
    }
    states = next; costs = nextCosts; paths = nextPaths;
  }

  const winner = costs.indexOf(Math.min(...costs));
  return paths[winner].map((k, i) => k === MISS ? null : perFrame[i][k]);
}

/**
 * Fill frames the detector missed *between* two good ones. Over the ~0.3s of a
 * clip the ball's path is close enough to straight for this to beat having no
 * reading at all. We deliberately do not extrapolate past the ends of the
 * track — that would be a guess, not an interpolation — and everything filled
 * in is flagged so the UI can say where the number came from.
 */
function fillGaps(track) {
  const known = [];
  track.forEach((b, i) => { if (b) known.push(i); });
  if (!known.length) return track.map(() => null);

  return track.map((b, i) => {
    if (b) return { ...b, interpolated: false };
    let before = null, after = null;
    for (const k of known) {
      if (k < i) before = k;
      else if (after === null) after = k;
    }
    if (before === null || after === null) return null;
    const t = (i - before) / (after - before);
    const a = track[before], c = track[after];
    return {
      x: a.x + (c.x - a.x) * t,
      y: a.y + (c.y - a.y) * t,
      r: a.r + (c.r - a.r) * t,
      score: Math.min(a.score, c.score),
      interpolated: true
    };
  });
}

/** Release wasm resources. */
export function closeBall() {
  detector?.close();
  detector = null;
  initPromise = null;
  clockMs = 0;
}

File: server/ball_detection.py
from __future__ import annotations

import math
import os

import cv2
import numpy as np

import models_cache
from biomechanics import LM, X, Y, V

BALL_CLASS_ID = 32
BALL_CATEGORY = "sports ball"

SCORE_FLOOR = float(os.environ.get("SFAI_BALL_SCORE_FLOOR", "0.15"))

IMGSZ = int(os.environ.get("SFAI_BALL_IMGSZ", "640"))

MIN_ROUNDNESS = 0.5
MIN_DIAM_TORSO = 0.18
MAX_DIAM_TORSO = 0.95

MISS_COST = 0.8
STEP_WEIGHT = 0.6
MAX_STEP_TORSO = 2.5


def _thread_budget() -> int:
    workers = max(1, int(os.environ.get("SFAI_WORKERS", "3")))
    return max(1, (os.cpu_count() or 4) // workers)


class _YoloBackend:

    def __init__(self, model_path: str | None = None):
        import torch
        from ultralytics import YOLO
        from ultralytics.utils import SETTINGS

        try:
            SETTINGS.update({"sync": False})
        except Exception:
            pass

        torch.set_num_threads(_thread_budget())
        self.model = YOLO(model_path or models_cache.ensure(models_cache.YOLO11X))

    def raw_boxes(self, images: list[np.ndarray]) -> list[list[dict]]:
        results = self.model.predict(images, classes=[BALL_CLASS_ID],
                                     conf=SCORE_FLOOR, imgsz=IMGSZ, verbose=False)
        out = []
        for r in results:
            frame = []
            for b in r.boxes:
                x1, y1, x2, y2 = (float(v) for v in b.xyxy[0].tolist())
                frame.append({"x": (x1 + x2) / 2, "y": (y1 + y2) / 2,
                              "w": x2 - x1, "h": y2 - y1, "score": float(b.conf[0])})
            out.append(frame)
        return out

    def close(self):
        pass


class _EfficientDetBackend:

    def __init__(self, model_path: str | None = None):
        import mediapipe as mp
        from mediapipe.tasks import python as mp_python
        from mediapipe.tasks.python import vision
        self._mp = mp

        path = model_path or models_cache.ensure(models_cache.DETECTOR_LITE2)
        self.detector = vision.ObjectDetector.create_from_options(
            vision.ObjectDetectorOptions(
                base_options=mp_python.BaseOptions(model_asset_path=path),
                running_mode=vision.RunningMode.IMAGE,
                category_allowlist=[BALL_CATEGORY],
                score_threshold=SCORE_FLOOR,
                max_results=8,
            ))

    def raw_boxes(self, images: list[np.ndarray]) -> list[list[dict]]:
        out = []
        for bgr in images:
            rgb = cv2.cvtColor(bgr, cv2.COLOR_BGR2RGB)
            image = self._mp.Image(image_format=self._mp.ImageFormat.SRGB, data=rgb)
            frame = []
            for d in getattr(self.detector.detect(image), "detections", None) or []:
                b = d.bounding_box
                if b is None or b.width <= 0 or b.height <= 0:
                    continue
                frame.append({"x": b.origin_x + b.width / 2, "y": b.origin_y + b.height / 2,
                              "w": float(b.width), "h": float(b.height),
                              "score": float(d.categories[0].score if d.categories else 0.0)})
            out.append(frame)
        return out

    def close(self):
        try:
            self.detector.close()
        except Exception:
            pass


class BallSession:

    def __init__(self, model_path: str | None = None):
        backend = os.environ.get("SFAI_BALL_BACKEND", "yolo").strip().lower()
        self.backend_name = "efficientdet" if backend == "efficientdet" else "yolo11x"
        self._backend = (_EfficientDetBackend(model_path)
                         if self.backend_name == "efficientdet"
                         else _YoloBackend(model_path))

    def close(self):
        self._backend.close()

    def __enter__(self):
        return self

    def __exit__(self, *exc):
        self.close()

    def detect_sequence(self, frames: list[dict], scales, all_pts=None) -> list:
        n = len(frames)
        filled = _fill_scales(scales, n)

        raw = self._backend.raw_boxes([f["image"] for f in frames])
        per_frame = [_candidates(raw[i], filled[i]) for i in range(n)]

        for i in range(n):
            if not per_frame[i]:
                pts = all_pts[i] if all_pts is not None else None
                per_frame[i] = self._rescue(frames[i]["image"], pts, filled[i])

        return _fill_gaps(_best_track(per_frame, filled))

    def _rescue(self, bgr, pts, scale):
        if pts is None or scale <= 0:
            return []
        h, w = bgr.shape[:2]
        feet = [pts[i] for i in (LM.L_AN, LM.R_AN, LM.L_FOOT, LM.R_FOOT) if pts[i][V] >= 0.3]
        if not feet:
            return []

        cx = float(np.mean([p[X] for p in feet]))
        cy = float(np.mean([p[Y] for p in feet]))
        pad = scale * 1.8
        x0, y0 = max(0, int(cx - pad)), max(0, int(cy - pad))
        x1, y1 = min(w, int(cx + pad)), min(h, int(cy + pad))
        if x1 - x0 < 16 or y1 - y0 < 16:
            return []

        crop = bgr[y0:y1, x0:x1]
        k = min(4.0, IMGSZ / max(crop.shape[0], crop.shape[1]))
        if k > 1.05:
            crop = cv2.resize(crop, None, fx=k, fy=k, interpolation=cv2.INTER_CUBIC)
        else:
            k = 1.0

        cands = _candidates(self._backend.raw_boxes([crop])[0], scale * k)
        for c in cands:
            c["x"] = c["x"] / k + x0
            c["y"] = c["y"] / k + y0
            c["r"] = c["r"] / k
            c["rescued"] = True
        return cands


def _fill_scales(scales, n):
    known = [s for s in (scales or []) if s is not None and math.isfinite(s) and s > 0]
    median = sorted(known)[len(known) // 2] if known else 0.0
    out = []
    for i in range(n):
        s = scales[i] if scales is not None and i < len(scales) else None
        out.append(s if (s is not None and math.isfinite(s) and s > 0) else median)
    return out


def _candidates(boxes, scale):
    out = []
    for b in boxes or []:
        w, h = b["w"], b["h"]
        if w <= 0 or h <= 0:
            continue
        if min(w, h) / max(w, h) < MIN_ROUNDNESS:
            continue

        diam = (w + h) / 2
        if scale and scale > 0:
            rel = diam / scale
            if rel < MIN_DIAM_TORSO or rel > MAX_DIAM_TORSO:
                continue

        out.append({"x": b["x"], "y": b["y"], "r": diam / 2,
                    "score": float(b["score"]), "rescued": False})
    return out


def _best_track(per_frame, scales):
    n = len(per_frame)
    if n == 0:
        return []
    MISS = -1

    def states_at(i):
        return list(range(len(per_frame[i]))) + [MISS]

    def emission(i, k):
        return MISS_COST if k == MISS else 1 - per_frame[i][k]["score"]

    def transition(i, frm, to):
        if frm == MISS or to == MISS:
            return 0.0
        a, b = per_frame[i - 1][frm], per_frame[i][to]
        s = scales[i] or scales[i - 1] or 0
        if not s:
            return 0.0
        step = math.hypot(a["x"] - b["x"], a["y"] - b["y"]) / s
        return STEP_WEIGHT * min(step, MAX_STEP_TORSO)

    states = states_at(0)
    costs = [emission(0, k) for k in states]
    paths = [[k] for k in states]

    for i in range(1, n):
        nxt = states_at(i)
        new_costs, new_paths = [], []
        for to in nxt:
            best, best_j = math.inf, 0
            for j, frm in enumerate(states):
                c = costs[j] + transition(i, frm, to)
                if c < best:
                    best, best_j = c, j
            new_costs.append(best + emission(i, to))
            new_paths.append(paths[best_j] + [to])
        states, costs, paths = nxt, new_costs, new_paths

    winner = costs.index(min(costs))
    return [None if k == MISS else per_frame[i][k] for i, k in enumerate(paths[winner])]


def _fill_gaps(track):
    known = [i for i, b in enumerate(track) if b]
    if not known:
        return [None] * len(track)

    out = []
    for i, b in enumerate(track):
        if b:
            out.append({**b, "interpolated": False})
            continue
        before = max((k for k in known if k < i), default=None)
        after = min((k for k in known if k > i), default=None)
        if before is None or after is None:
            out.append(None)
            continue
        t = (i - before) / (after - before)
        a, c = track[before], track[after]
        out.append({
            "x": a["x"] + (c["x"] - a["x"]) * t,
            "y": a["y"] + (c["y"] - a["y"]) * t,
            "r": a["r"] + (c["r"] - a["r"]) * t,
            "score": min(a["score"], c["score"]),
            "rescued": False,
            "interpolated": True,
        })
    return out
